ray_lightning/ray_ddp.py

```diff
@@ -176,19 +167,58 @@ def _setup_env_vars(self):
         values = [os.getenv(k) for k in keys]
         ray.get([w.set_env_vars.remote(keys, values) for w in self.workers])
 
-    def execution_loop(self, trainer, tune_enabled: bool = True):
+    def start_training(self, trainer):
+        results = self.execution_loop(tune_enabled=True)
+        # reset optimizers, since main process is never used for training and
+        # thus does not have a valid optim state.
+        trainer.optimizers = []
+        return results
+
+    def start_evaluating(self, trainer):
+        results = self.execution_loop(tune_enabled=False)
+        return results
+
+    def start_predicting(self, trainer):
+        results = self.execution_loop(tune_enabled=False)
+        return results
+
+    def get_local_ranks(self) -> Dict[int, int]:
+        """Creates a mapping of global ranks to local ranks/node ranks."""
+        # Get the local ranks for all the workers and store as a dict.
+        # First get the IP address of each remote worker.
+        node_ips = ray.get([w.get_node_ip.remote() for w in self.workers])
+
+        node_ip_map = {}
+        for i in range(len(node_ips)):
+            node_ip_map[node_ips[i]] = i
+
+        rank_counter_dict = defaultdict(int)
+        global_to_local = [None] * self.num_workers
+
+        for global_rank in range(self.num_workers):
+            ip = node_ips[global_rank]
+            global_to_local[global_rank] = (rank_counter_dict[ip],
+                                            node_ip_map[ip])
+            rank_counter_dict[ip] += 1
```

Comment:
If I'm reading this correctly, the second item in the value tuple (`node_ip_map[ip]`) should be the node rank, but from this logic it looks like it's being set to the `global_rank`?